core/build.gradle

```
@@ -74,6 +75,7 @@ dependencies {
     api ('org.rnorth.duct-tape:duct-tape:1.0.8') {
         exclude(group: 'org.jetbrains', module: 'annotations')
     }
+    implementation 'com.google.cloud.tools:jib-core:0.22.0'
```

To avoid a direct dependency on Jib:
```suggestion
    implementationOnly 'com.google.cloud.tools:jib-core:0.22.0'
```

---

with `implementation` there is no direct dependency, unlike `api`. User should add the dependency explicitly along with testcontainers in order to make it work as suggested by the docs. Am I missing something?

---

I think Maven users would still get the transitive dependency, since Maven does not know about the distinction between `api` and `implementation`. At least I used to believe this, is it still correct?

---

the pom.xml produced doesn't container any entry for jib.

---

Something I found while looking into the best practices around this, could we explore using Gradle's feature variant capabilities for this, and it should provide the best experience for Maven and for Gradle users?
https://docs.gradle.org/current/userguide/feature_variants.html

---

Unfortunately, we can't use feature variants with our current Gradle build and publication code (something to look into in the future). We looked at how `implementation` currently behaves for Gradle and Maven consumers and this means we won't have `jib` in the published POM and hence consumers won't get this dependency automatically (double checked this with local build).

Gradle themselves also recommend against using `compileOnly` for modelling optional dependencies. 

We think for now it's ok to go ahead with `implementation` and consider the feature variants approach in the future.

---

1. let's not use any Gradle-specific features, or otherwise we will have "fun" time dealing with user reports where the behavior will be dramatically different if they are coming from Maven or Gradle 😅
2. `implementation` is considered a transitive non-API dependency, "implementation detail", while we want to _link_ against it, but it is not needed for TC to work.
3. I just remembered that we have `provided` scope configured ([including publishing](https://github.com/testcontainers/testcontainers-java/blob/65b6aeb40922ea9d0e5e6fa9e607cfbde24512ae/gradle/publishing.gradle#L87)) for this particular use case: https://github.com/testcontainers/testcontainers-java/blob/65b6aeb40922ea9d0e5e6fa9e607cfbde24512ae/modules/selenium/build.gradle#L6

---

1. I can understand this concern and the effort of staying up-to-date with changes around those features. Dealing with user issues can indeed be challenging if we are not deep in the mental model of Gradle (and we already saw while looking into this topic, that it can be a rabbit hole).
2. True, regarding Gradle semantics, `implementation` is not correct for achieving `provided`
3. Right 😅 Wasn't aware of this anymore (which I think also points at us trying to remove complexity of the Gradle build rather than increasing it). So using `provided` here seems straight forward and consistent within the context of our build. I'd say we should use it here, WDYT @eddumelendez ?

---

@kiview already applied and merged 😄